### ISSUE TITLE
project overlap / parser separation

### ISSUE THREAD

**Issue by smhg** [zaach/jsxgettext#82]
I've been working on [xgettext-template](https://github.com/gmarty/xgettext) for a while, which seems to be very similar to this project.
I haven't looked into things enough to see what's different (besides language support), so I'm avoiding the overlap-subject for now.

I came here with this question: since I was working on Swig-parsing, I wondered if you would consider splitting up the parser(s) into (a) different sub-project(s)?
xgettext-template was recently restructured with that in mind: the [Handlebars parser](https://github.com/smhg/gettext-handlebars) is a separate project that can be used independently from the main project (basically the CLI interface). Definitely feel free to use it (it doesn't rely on regex's, but on the Handlebars lib itself).

**Comment by beck** [comment]
I just started diving into this project have a bit to add:

> use the template language's parser/lexer if possible

:+1: 

> I think a parser needs to accept a keyword spec ... template string. 

There's already an implicit API for how a parser is interfaced, two arguments are sent:

1. `sources` - a dictionary of `{"file1.js": "content", "file2.js": "content", ... }`
1. `options` - all config passed in via the command line (including `keyword`)

I think this is a good starting point for CLI > Parser communication.

As for the response, this is where this project needs some improvement:

Currently the response is:

1. `sources`: a dictionary of filename / **javascript source**
1. `options`

I propose that the `sources` dictionary does not return javascript but a dictionary of translations [as spec'd by the gettext-parser](https://github.com/andris9/gettext-parser#translations).

I think this would yield significant cleanup, including:
* Javascript parsing is factored out of `jsxgettext` and moved into it's own `javascript.js` under `lib/parsers/`
* The responsibility of `jsxgettext.js` become more precise: interfacing with parsers and `cli.js`
* Files would no longer be parsed twice (once as their own language, then again as js)
* Parsers become simplified as they no longer have to coerce messages into valid js gettext arguments

> Updating the CLI package every time a new parser is added or a new version is released, feels so troublesome

True, but it is not too bad.

If all are in agreement about proposed changes, then focus should be on refactoring and developing the internal API.  Once that's done, then focus can be on user extensibility with installable parsers.

**Comment by BYK** [comment]
>>Updating the CLI package every time a new parser is added or a new version is released, feels so troublesome
>True, but it is not too bad.

Yup. Moreover we can add some hooks for new parsers to attach themselves (or make them discoverable) dynamically.

>If all are in agreement about proposed changes, then focus should be on refactoring and developing the internal API. Once that's done, then focus can be on user extensibility with installable parsers.

Heh, I wrote the thing above before reading this. Yeah I think this sounds like a good path forward.

**Comment by smhg** [comment]
Good suggestion about using [gettext-parser's structure](https://github.com/andris9/gettext-parser#translations).
However, this, combined with sending a dictionary of file contents, makes each parser responsible for result aggregation.
Conceptually that sounds more like a job for the "parent" (cli) package, no?

poEdit calls will probably always be for one (template) language so no aggregation should be performed across languages. But you could have calls where the language is derived from the file extension and thus have cases where you need to aggregate the results of multiple parsers. Not common for poEdit, but it might make more sense for something like grunt.

In that light, you could argue a parser just needs to focus on parsing one input string. The result would of course resemble gettext-parser's structure as close as possible.

**Comment by BYK** [comment]
>However, this, combined with sending a dictionary of file contents, makes each parser responsible for result aggregation.
>Conceptually that sounds more like a job for the "parent" (cli) package, no?

I agree with this one.

>poEdit calls will probably always be for one (template) language so no aggregation should be performed across languages.

Sort of disagree. I think we should always do aggregation but should not expect it from individual parsers. May be unless it is in the same file? Not sure. I want to make parsers as simple as possible and gettext-parser's structure has a bit more than that. On the flip side, if we do not enforce this structure we would end up inventing our own structure which would be quite similar to this one I think.

**Comment by beck** [comment]
>  a parser just needs to focus on parsing one input string

:+1:

**Comment by smhg** [comment]
Great. Let's agree on parsers' use and its output format then.
To get a discussion started, gettext-handlebars is used like this:

```javascript
var parser = new Parser({
  _: [0],
  gettext: [0],
  ngettext: [0, 1]
});

parser.parse(contentOfOneFile);
```
* keywordspec is passed to constructor (can there be other options?)
* exposes a `parse` method that takes a string

Please adapt the above as you see fit and feel free to add a sample return value format.

**Comment by beck** [comment]
I would prefer that parsers be instantiated with convention > configuration.

For example, if the handlebars are already using `gettext` and `ngettext` then the parser should not need any additional config. (side note: I've stopped using `_` as good convention because it too often collides with those using lodash, underscore, or as a placeholder for throwaways).

Question: why is the argument array necessary?  Are there cases where a keyword func would get anything other than one argument (simple translation) or three (plural)?

If one needs to override/extend the agreed-upon convention, (say the handlebars use `trans` or `_` instead of `gettext`) then this should be provided with the `keyword` option.

I think the parser should be provided all options (so when something like `debug` is added to the cli, that information is automatically provided to all parsers).  I would recommend:

```
var parser = new Parser(options);
var translations = parser.parse(contentOfOneFile);
```

**Comment by smhg** [comment]
I'm definitely in favour of providing `options`.
Even if there isn't anything else besides a keywordspec which is ever passed.
Let's agree on the full option names of [xgettext](https://www.gnu.org/savannah-checkouts/gnu/gettext/manual/html_node/xgettext-Invocation.html) as property names for the options object? With dashes turned into camelcase? So the keywordspec would indeed be under `keyword`.

I'm not entirely sure about the defaults though. It is definitely true `xgettext` has defaults. I think these of course should be honoured. I mean: at least in the cli package.
What I'm not sure about is whether every parser needs to have defaults. Parsers can be quite small so this would add a _relatively_ high amount of duplicate code.
On the other hand: for stand alone usage of parsers default keywords are nice.
Since this is a rather minor thing: maybe we go for defaults in every parser and reevaluate this later?

About the argument position array for every keyword: parsers need to accept these. Yes, people (including myself) use others than the defaults. In general: I would always try to offer full xgettext compatibility.
We could however additionally handle a flat array of keywords too if you prefer. It's related to the above though: do we want ever parser to handle this on its own or not?

**Comment by ArmorDarks** [comment]
Not much I can help here, but just for the sake of at least some support — I think it will be great feature. Hope to see it in nearby future.
